kpi/serializers/v2/asset_permission.py

```diff
@@ -0,0 +1,209 @@
+        user = validated_data['user']
+        asset = validated_data['asset']
+        permission = validated_data['permission']
+        partial_permissions = validated_data.get('partial_permissions', None)
+        return asset.assign_perm(user, permission.codename,
+                                 partial_perms=partial_permissions)
+
+    def get_partial_permissions(self, object_permission):
+        codename = object_permission.permission.codename
+        if codename.startswith(PREFIX_PARTIAL_PERMS):
+            view = self.context.get('view')
+            # if view doesn't have an `asset` property,
+            # fallback to context. (e.g. AssetViewSet)
+            asset = getattr(view, 'asset', self.context.get('asset'))
+            partial_perms = asset.get_partial_perms(
+                object_permission.user_id, True)
+
+            hyperlinked_partial_perms = []
+            for perm_codename, filters in partial_perms.items():
+                url = self.__get_permission_hyperlink(perm_codename)
+                hyperlinked_partial_perms.append({
+                    'url': url,
+                    'filters': filters
+                })
+            return hyperlinked_partial_perms
+        return None
+
+    def get_url(self, object_permission):
+        asset_uid = self.context.get('asset_uid')
+        return reverse('asset-permission-detail',
+                       args=(asset_uid, object_permission.uid),
+                       request=self.context.get('request', None))
+
+    def validate(self, attrs):
+        # Because `partial_permissions` is a `SerializerMethodField`,
+        # it's read-only, so it's not validated nor added to `validated_data`.
+        # We need to do it manually
+        self.validate_partial_permissions(attrs)
+        return attrs
+
+    def validate_partial_permissions(self, attrs):
+        """
+        Validates permissions and filters sent with partial permissions.
+
+        If data is valid, `partial_permissions` attribute is added to `attrs`.
+        Useful to permission assignment in `create()`.
+
+        :param attrs: dict.
+        :return: dict.
+        """
+        permission = attrs.get('permission')
+        request = self.context.get('request')
+        if isinstance(request.data, dict):
+            partial_permissions = request.data.get('partial_permissions')
+        elif self.context.get('partial_permissions'):
+            partial_permissions = self.context.get('partial_permissions')
+        partial_permissions_attr = {}
+
+        if permission.codename.startswith(PREFIX_PARTIAL_PERMS):
+            if partial_permissions:
+                is_valid = True
+                try:
+                    for partial_permission, filter_ in \
+                            self.__get_partial_permissions_generator(partial_permissions):
+
+                        parse_result = urlparse(partial_permission.get('url'))
+                        resolver = resolve(parse_result.path)
```

Comment:
@jnm, 
`partial_permission.get('url')` equals None, `urlparse(partial_permission.get('url'))` will raise an `AttributeError`.

I've narrowed down the try/except block, but I've kept the `AttributeError` in case.

```python
                    try:
                        parse_result = urlparse(partial_permission.get('url'))
                        resolver = resolve(parse_result.path)
                        codename = resolver.kwargs.get('codename')
                    except (AttributeError, Resolver404):
                        is_valid = False
                        break  # No need to going further
```
I've also added a `break` in the loop just after `is_valid = False`, because if one permission is not valid, no need to evaluate others. What do you think?

Reply:
We gave this its due (and then some :laughing:)